biz.aQute.tester.junit-platform/src/aQute/tester/bundle/engine/discovery/BundleSelectorResolver.java

```diff
@@ -398,6 +401,51 @@ private void markClassesResolved(Bundle bundle) {
 		}
 	}
 
+	private void checkForMixedJUnit34(BundleDescriptor bd, Class<?> testClass) {
+		try {
+			// Don't hardcode TestCase.class here as we don't want to introduce
+			// a hard dependency on that package
+			Class<?> junit3TestCase = testClass.getClassLoader()
+				.loadClass("junit.framework.TestCase");
+
+			if (junit3TestCase.isAssignableFrom(testClass) && hasJUnit4Annotations(testClass)) {
+				StaticFailureDescriptor mixedError = new StaticFailureDescriptor(bd.getUniqueId()
+					.append("test", testClass.getName()), testClass.getName(),
+					new JUnitException(
+						"Class extends junit.framework.TestCase and has JUnit 4 annotations; annotations will be ignored"));
+				bd.addChild(mixedError);
+			}
+
+		} catch (ClassNotFoundException e) {
+		}
+	}
+
+	private boolean hasJUnit4Annotations(Class<?> clazz) {
+		if (elementHasJUnit4Annotations(clazz))
+			return true;
+
+		for (Method m : clazz.getMethods()) {
+			if (elementHasJUnit4Annotations(m)) {
+				return true;
+			}
+		}
+		return false;
+	}
+
+	private boolean elementHasJUnit4Annotations(AnnotatedElement element) {
+		Annotation[] annotations = element.getAnnotations();
+		if (annotations != null) {
+			for (Annotation a : annotations) {
+				final String annotationName = a.annotationType()
+					.getName();
+				if (annotationName.startsWith("org.junit.") && !annotationName.contains(".jupiter.")) {
```

Comment:
I wonder if this is focused enough in JUnit 4? Perhaps some future JUnit 6 will have annotations which start with `org.junit.` but do not contain `.jupiter.`. Then we will be mistaken. Perhaps we need a list of JUnit 4 annotation names in a collection and can ask that collection if it contains the observed annotation name. The doesn't-contain test is not future proof enough.

Reply:
Fair point - the latest push has gone with a fairly exhaustive check of specifically JUnit 4 annotations rather than a generic pattern.